### PR TITLE
partial accumulator atomic model

### PR DESCRIPTION
**Pull request by cruizm**

### CODE REVIEW

File: include/cadmium/basic_model/partial_accumulator.hpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    /**
+     * @brief PartialAccumulator PDEVS Model.
+     *
+     * PartialAccumulator PDEVS Model:
+     * - X = Numeric
+     * - Y = Numeric
+     * - In_Ports: add<Numeric>, reset, partial
+     * - Out_Ports: outport
+     * - S = {Numeric:total, bool:reseted, int:partial}
```
Comment (sdavtaker) [inline include/cadmium/basic_model/partial_accumulator.hpp]: Why Partial is an integer? Shouldn't it also be a bool as reset in S?

File: include/cadmium/basic_model/partial_accumulator.hpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+     * - internal({total, true, i}) = {0, false, 0}
+     *   internal({total, false, 1}) = {total, false, 0}
+     * - external({total, b, i}, t, x.add) = {total+x, b, i}
+     *   external({total, b, i}, t, x.reset) = {total, true, i}
+     *   external({total, b, i}, t, x.partial) = {total, b, 1}
+     * - confluence({total, b, i}, 0, x) = external(internal({total, b, i}), 0, x)
```
Comment (sdavtaker) [inline include/cadmium/basic_model/partial_accumulator.hpp]: External definition can receive Bags of events, for example Reset and Partial.
In that case I will prefer to partial first and reset second, but it is not defined here.

File: include/cadmium/basic_model/partial_accumulator.hpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+     * - X = Numeric
+     * - Y = Numeric
+     * - In_Ports: add<Numeric>, reset, partial
+     * - Out_Ports: outport
+     * - S = {Numeric:total, bool:reseted, int:partial}
+     * - internal({total, true, i}) = {0, false, 0}
+     *   internal({total, false, 1}) = {total, false, 0}
+     * - external({total, b, i}, t, x.add) = {total+x, b, i}
+     *   external({total, b, i}, t, x.reset) = {total, true, i}
+     *   external({total, b, i}, t, x.partial) = {total, b, 1}
```
Comment (sdavtaker) [inline include/cadmium/basic_model/partial_accumulator.hpp]: External definition can receive Bags of events, for example Reset and Partial.
In that case I will prefer to partial first and reset second, but it is not defined here.

File: include/cadmium/basic_model/partial_accumulator.hpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        };
+        struct add : public in_port<VALUE> {
+        };
+        struct reset : public in_port<SET> {
+        };
+        struct partial : public in_port<SET>{
+        };
+    };   
+
+    template<typename VALUE, typename SET, typename TIME> //value is the type of accumulated values
```
Comment (sdavtaker) [inline include/cadmium/basic_model/partial_accumulator.hpp]: A comment about what is SET would be nice

File: include/cadmium/basic_model/partial_accumulator.hpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      struct partial_accumulator {
+        using on_reset=bool;
+        using on_partial=int;
```
Comment (sdavtaker) [inline include/cadmium/basic_model/partial_accumulator.hpp]: Maybe you want to define the state as :
template<typename VALUE>
struct state_type{ 
    bool on_reset;
    bool on_partial;
    VALUE accumulated;
};

File: include/cadmium/basic_model/partial_accumulator.hpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        //default constructor
+        constexpr partial_accumulator() noexcept {}
+
+        //ports_definition
+        using input_ports=std::tuple<typename defs::add, typename defs::reset, typename defs::partial>;
+        using output_ports=std::tuple<typename defs::sum>;
+
+        // PDEVS functions
+        void internal_transition() {
+          if(!std::get<on_reset>(state) && std::get<on_partial>(state)!=1) {
```
Comment (sdavtaker) [inline include/cadmium/basic_model/partial_accumulator.hpp]: I would change the state.on_partial type to bool as mentioned before

File: include/cadmium/basic_model/partial_accumulator.hpp [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            throw std::logic_error("Internal transition called while not on reset or not on partial state");
+          }
+          if(std::get<on_reset>(state)) {
```
Comment (sdavtaker) [inline include/cadmium/basic_model/partial_accumulator.hpp]: If you make the change to use bool, in place of using get, you should just call the variable inside the struct:
if (std::get<on_reset>(state))
should be
if (state.on_reset)

File: test/pdevs_partial_accumulator_model_test.cpp [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+            BOOST_CHECK_EQUAL(0, std::get<int>(g.state));
+            
+             //validate output
+            typename cadmium::make_message_bags<floating_partial_accumulator::output_ports>::type outmb2;
+            outmb2 = g.output();
+            BOOST_CHECK_EQUAL(true, cadmium::get_messages<typename floating_partial_accumulator_defs::sum>(outmb2).empty());
+        }
+
+    BOOST_AUTO_TEST_SUITE_END()
+
```
Comment (sdavtaker) [inline test/pdevs_partial_accumulator_model_test.cpp:236]: I would like to have a test of what happens when you sent input in the 3 ports or pairs of ports at the same time. In special, reset and partial ports.

File: include/cadmium/basic_model/partial_accumulator.hpp (hunk 1 above)
Comment (cruizm) [reply include/cadmium/basic_model/partial_accumulator.hpp]: You are right. It doesn't make any sense. I did that to have the state as a tuple. However, I totally agree. I prefer the state as a struct with bool partial.

File: include/cadmium/basic_model/partial_accumulator.hpp (hunk 3 above)
Comment (cruizm) [reply include/cadmium/basic_model/partial_accumulator.hpp]: I will define the multiple combinations of the messages in the external

File: test/pdevs_partial_accumulator_model_test.cpp (hunk 8 above)
Comment (cruizm) [reply test/pdevs_partial_accumulator_model_test.cpp:236]: I will add the test

Comment (cruizm) [reply test/pdevs_partial_accumulator_model_test.cpp:236]: Some of them were already done, I have added more combinations. They are inside  BOOST_AUTO_TEST_CASE( output_function_returns_accumulated_value_test_and_external_combinations )

**Pull request closed without merging**